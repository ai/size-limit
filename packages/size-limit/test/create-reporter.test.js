let createReporter = require('../create-reporter')

function results(types, config, isJSON = false, isSilentMode = false) {
  let stdout = ''
  let plugins = {
    has(type) {
      return types.includes(type)
    }
  }
  let process = {
    stdout: {
      write(str) {
        stdout += str
      }
    }
  }
  let reporter = createReporter(process, isJSON, isSilentMode)
  reporter.results(plugins, config)
  return stdout
}

it('renders results', () => {
  expect(
    results(['webpack', 'time'], {
      checks: [
        {
          name: 'limitless',
          size: 10,
          config: 'a',
          loadTime: 0.1,
          runTime: 0.5,
          time: 0.6
        },
        {
          name: 'size',
          size: 102400,
          sizeLimit: 102400,
          loadTime: 1,
          runTime: 2,
          time: 3,
          passed: true
        },
        {
          name: 'time',
          size: 102400,
          gzip: false,
          timeLimit: 4,
          loadTime: 1,
          runTime: 2,
          time: 3,
          passed: true
        }
      ]
    })
  ).toMatchSnapshot()
})

it('renders list of success checks in silent mode', () => {
  expect(
    results(
      ['webpack', 'time'],
      {
        checks: [
          {
            name: 'limitless',
            size: 10,
            config: 'a',
            loadTime: 0.1,
            runTime: 0.5,
            time: 0.6
          },
          {
            name: 'size',
            size: 102400,
            sizeLimit: 102400,
            loadTime: 1,
            runTime: 2,
            time: 3,
            passed: true
          }
        ]
      },
      false,
      true
    )
  ).toMatchSnapshot()
})

it('renders list of failed checks in silent mode', () => {
  expect(
    results(
      ['webpack', 'time'],
      {
        checks: [
          {
            name: 'small fail',
            size: 102401,
            sizeLimit: 102400,
            passed: false
          },
          {
            name: 'big fail',
            size: 102500,
            sizeLimit: 102400,
            passed: false
          }
        ]
      },
      false,
      true
    )
  ).toMatchSnapshot()
})

it('renders list of failed and success checks in silent mode', () => {
  expect(
    results(
      ['webpack', 'time'],
      {
        checks: [
          {
            name: 'small fail',
            size: 102401,
            sizeLimit: 102400,
            passed: false
          },
          {
            name: 'limitless',
            size: 10,
            config: 'a',
            loadTime: 0.1,
            runTime: 0.5,
            time: 0.6
          },
          {
            name: 'size',
            size: 102400,
            sizeLimit: 102400,
            loadTime: 1,
            runTime: 2,
            time: 3,
            passed: true
          },
          {
            name: 'big fail',
            size: 102500,
            sizeLimit: 102400,
            passed: false
          }
        ]
      },
      false,
      true
    )
  ).toMatchSnapshot()
})

it('renders failed results', () => {
  expect(
    results(['file'], {
      checks: [
        {
          name: 'ok',
          size: 102400,
          sizeLimit: 102400,
          passed: true
        },
        {
          name: 'small fail',
          size: 102401,
          sizeLimit: 102400,
          passed: false
        },
        {
          name: 'big fail',
          size: 102500,
          sizeLimit: 102400,
          passed: false
        }
      ],
      failed: true,
      configPath: 'package.json'
    })
  ).toMatchSnapshot()
})

it('only renders failed results with --hide-passed flag', () => {
  expect(
    results(['file'], {
      checks: [
        {
          name: 'ok',
          size: 102400,
          sizeLimit: 102400,
          passed: true
        },
        {
          name: 'small fail',
          size: 102401,
          sizeLimit: 102400,
          passed: false
        },
        {
          name: 'big fail',
          size: 102500,
          sizeLimit: 102400,
          passed: false
        }
      ],
      hidePassed: true,
      failed: true,
      configPath: 'package.json'
    })
  ).toMatchSnapshot()
})

it('highlight passed results with --highlight-less flag', () => {
  expect(
    results(['file'], {
      checks: [
        {
          name: 'ok',
          size: 102400,
          sizeLimit: 102400,
          passed: true
        },
        {
          name: 'good',
          size: 97280, // 5 KiB less
          sizeLimit: 102400,
          passed: true
        },
        {
          name: 'fail',
          size: 102401,
          sizeLimit: 102400,
          passed: false
        }
      ],
      highlightLess: true,
      failed: true,
      configPath: 'package.json'
    })
  ).toMatchSnapshot()
})

it('renders single result', () => {
  expect(
    results(['file'], {
      checks: [
        {
          name: 'big fail',
          size: 101,
          sizeLimit: 100,
          passed: false
        }
      ],
      failed: true,
      configPath: '.size-limit.json'
    })
  ).toMatchSnapshot()
})

it('renders config-less result', () => {
  expect(
    results(['time'], {
      checks: [
        {
          name: 'big fail',
          size: 1000,
          timeLimit: 0.5,
          loadTime: 0.2,
          runTime: 0.3,
          time: 0.5,
          passed: false
        }
      ],
      failed: true
    })
  ).toMatchSnapshot()
})

it('renders JSON results', () => {
  expect(
    results(
      ['file'],
      {
        checks: [
          {
            name: 'big fail',
            path: '/a'
          },
          {
            name: 'big fail',
            size: 1000,
            timeLimit: 10,
            loadTime: 0.2,
            runTime: 0.3,
            time: 0.5,
            passed: false,
            path: '/b'
          }
        ],
        failed: true
      },
      true
    )
  ).toMatchSnapshot()
})

it('renders result for file without gzip', () => {
  expect(
    results(['file'], {
      checks: [
        {
          name: 'without gzip',
          size: 100,
          sizeLimit: 99,
          passed: true,
          gzip: false
        }
      ],
      failed: false,
      configPath: '.size-limit.json'
    })
  ).toMatchSnapshot()
})

it('renders result for file with brotli', () => {
  expect(
    results(['file'], {
      checks: [
        {
          name: 'with brotli',
          size: 100,
          sizeLimit: 99,
          passed: true,
          brotli: true
        }
      ],
      failed: false,
      configPath: '.size-limit.json'
    })
  ).toMatchSnapshot()
})

it('renders Webpack stats help message', () => {
  expect(
    results(['webpack'], {
      checks: [],
      saveBundle: 'test'
    })
  ).toMatchSnapshot()
})
