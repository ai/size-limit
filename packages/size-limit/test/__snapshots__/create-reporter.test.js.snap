// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`highlight passed results with --highlight-less flag 1`] = `
"  
  [1mok[22m
  Size limit: [32m[1m102.4 kB[22m[39m
  Size:       [32m[1m102.4 kB[22m[39m [90mgzipped[39m
  
  [1mgood[22m
  Size limit: [32m[1m102.4 kB[22m[39m
  Size:       [32m[1m97.28 kB[22m[39m [90mgzipped[39m
  
  [1mfail[22m
  [31mPackage size limit has exceeded by 1 B[39m
  Size limit: [31m[1m102400 B[22m[39m
  Size:       [31m[1m102401 B[22m[39m [90mgzipped[39m
  
  [33mTry to reduce size or increase limit in [1m\"size-limit\"[22m section of [1mpackage.json[22m[39m
"
`;

exports[`only renders failed results with --hide-passed flag 1`] = `
"  
  [1msmall fail[22m
  [31mPackage size limit has exceeded by 1 B[39m
  Size limit: [31m[1m102400 B[22m[39m
  Size:       [31m[1m102401 B[22m[39m [90mgzipped[39m
  
  [1mbig fail[22m
  [31mPackage size limit has exceeded by 100 B[39m
  Size limit: [31m[1m102.4 kB[22m[39m
  Size:       [31m[1m102.5 kB[22m[39m [90mgzipped[39m
  
  [33mTry to reduce size or increase limit in [1m\"size-limit\"[22m section of [1mpackage.json[22m[39m
"
`;

exports[`renders JSON results 1`] = `
"[
  {
    "name": "big fail"
  },
  {
    "name": "big fail",
    "passed": false,
    "size": 1000,
    "running": 0.3,
    "loading": 0.2
  }
]
"
`;

exports[`renders Webpack stats help message 1`] = `
"  
  Webpack Stats file was saved to test/stats.json
  You can review it using https://webpack.github.io/analyse
"
`;

exports[`renders config-less result 1`] = `
"  
  [31mTotal time limit has exceeded[39m
  Time limit:   [31m[1m500 ms[22m[39m
  Size:         [31m[1m1 kB[22m[39m
  Loading time: [31m[1m200 ms[22m[39m [90mon slow 3G[39m
  Running time: [31m[1m300 ms[22m[39m [90mon Snapdragon 410[39m
  Total time:   [31m[1m500 ms[22m[39m
  
  [33mTry to reduce size or increase limit[39m
"
`;

exports[`renders failed results 1`] = `
"  
  [1mok[22m
  Size limit: [32m[1m102.4 kB[22m[39m
  Size:       [32m[1m102.4 kB[22m[39m [90mgzipped[39m
  
  [1msmall fail[22m
  [31mPackage size limit has exceeded by 1 B[39m
  Size limit: [31m[1m102400 B[22m[39m
  Size:       [31m[1m102401 B[22m[39m [90mgzipped[39m
  
  [1mbig fail[22m
  [31mPackage size limit has exceeded by 100 B[39m
  Size limit: [31m[1m102.4 kB[22m[39m
  Size:       [31m[1m102.5 kB[22m[39m [90mgzipped[39m
  
  [33mTry to reduce size or increase limit in [1m\"size-limit\"[22m section of [1mpackage.json[22m[39m
"
`;

exports[`renders list of failed and success checks in silent mode 1`] = `
"  
  [1msmall fail[22m
  [31mPackage size limit has exceeded by 1 B[39m
  Size limit: [31m[1m102400 B[22m[39m
  Size:       [31m[1m102401 B[22m[39m [90mwith all dependencies, minified and gzipped[39m
  
  [1mbig fail[22m
  [31mPackage size limit has exceeded by 100 B[39m
  Size limit: [31m[1m102.4 kB[22m[39m
  Size:       [31m[1m102.5 kB[22m[39m [90mwith all dependencies, minified and gzipped[39m
  
"
`;

exports[`renders list of failed checks in silent mode 1`] = `
"  
  [1msmall fail[22m
  [31mPackage size limit has exceeded by 1 B[39m
  Size limit: [31m[1m102400 B[22m[39m
  Size:       [31m[1m102401 B[22m[39m [90mwith all dependencies, minified and gzipped[39m
  
  [1mbig fail[22m
  [31mPackage size limit has exceeded by 100 B[39m
  Size limit: [31m[1m102.4 kB[22m[39m
  Size:       [31m[1m102.5 kB[22m[39m [90mwith all dependencies, minified and gzipped[39m
  
"
`;

exports[`renders list of success checks in silent mode 1`] = `
"  
"
`;

exports[`renders result for file with brotli 1`] = `
"  
  Size limit: [32m[1m99 B[22m[39m
  Size:       [32m[1m100 B[22m[39m [90mbrotlied[39m
  
"
`;

exports[`renders result for file without gzip 1`] = `
"  
  Size limit: [32m[1m99 B[22m[39m
  Size:       [32m[1m100 B[22m[39m
  
"
`;

exports[`renders results 1`] = `
"  
  [1mlimitless[22m
  Size:         [1m10 B  [22m [90mwith given webpack configuration[39m
  Loading time: [1m100 ms[22m [90mon slow 3G[39m
  Running time: [1m500 ms[22m [90mon Snapdragon 410[39m
  Total time:   [1m600 ms[22m
  
  [1msize[22m
  Size limit:   [32m[1m102.4 kB[22m[39m
  Size:         [32m[1m102.4 kB[22m[39m [90mwith all dependencies, minified and gzipped[39m
  Loading time: [32m[1m1 s     [22m[39m [90mon slow 3G[39m
  Running time: [32m[1m2 s     [22m[39m [90mon Snapdragon 410[39m
  Total time:   [32m[1m3 s[22m[39m
  
  [1mtime[22m
  Time limit:   [32m[1m4 s[22m[39m
  Size:         [32m[1m102.4 kB[22m[39m [90mwith all dependencies and minified[39m
  Loading time: [32m[1m1 s     [22m[39m [90mon slow 3G[39m
  Running time: [32m[1m2 s     [22m[39m [90mon Snapdragon 410[39m
  Total time:   [32m[1m3 s[22m[39m
  
"
`;

exports[`renders single result 1`] = `
"  
  [31mPackage size limit has exceeded by 1 B[39m
  Size limit: [31m[1m100 B[22m[39m
  Size:       [31m[1m101 B[22m[39m [90mgzipped[39m
  
  [33mTry to reduce size or increase limit at [1m.size-limit.json[22m[39m
"
`;
